training_data_loader.cpp

```diff
@@ -83,17 +84,20 @@ struct HalfKPFactorized {
     static constexpr int MAX_PIECE_FEATURES = 32;
     static constexpr int MAX_ACTIVE_FEATURES = HalfKP::MAX_ACTIVE_FEATURES + MAX_K_FEATURES + MAX_PIECE_FEATURES;
 
-    static void fill_features_sparse(int i, const TrainingDataEntry& e, int* features, int& counter, Color color)
+    static void fill_features_sparse(int i, const TrainingDataEntry& e, int* features, float* values, int& counter, Color color)
     {
-        int offset = HalfKP::fill_features_sparse(i, e, features, counter, color);
+        auto counter_before = counter;
+        int offset = HalfKP::fill_features_sparse(i, e, features, values, counter, color);
         auto& pos = e.pos;
         {
+            auto num_added_features = counter - counter_before;
             // king square factor
             auto ksq = pos.kingSquare(color);
             int idx = counter * 2;
-            counter += 1;
             features[idx] = i;
             features[idx + 1] = offset + static_cast<int>(orient(color, ksq));
+            values[counter] = static_cast<float>(num_added_features);
```

Comment:
This works because we don't include the kings as pieces, but perhaps calling the variable `num_non_king_pieces` or something might be a bit clearer?

Reply:
Even if the (king king) features are included (as in halfka) each such (king king) feature can be factorized the same all others. So this value is not `num_non_king_pieces`

Reply:
I see, yes, that's a good point.